src/Database/LSMTree/Internal/IndexOrdinary.hs

```diff
@@ -0,0 +1,222 @@
+-- | A general-purpose fence pointer index.
+module Database.LSMTree.Internal.IndexOrdinary
+(
+    IndexOrdinary (IndexOrdinary),
+
+    -- * Search
+    search,
+
+    -- * Deserialisation
+    fromSBS
+)
+where
+
+import Prelude hiding (length, drop, takeWhile)
+
+import Control.Monad
+       (
+           when
+       )
+import Data.Word
+       (
+           Word8,
+           Word16,
+           Word32,
+           Word64,
+           byteSwap32
+       )
+import Data.Primitive.ByteArray
+       (
+           ByteArray,
+           indexByteArray
+       )
+import Data.ByteString.Short
+       (
+           ShortByteString (ShortByteString)
+       )
+import Data.ByteString.Short qualified as ShortByteString
+       (
+           length
+       )
+import Data.Vector
+       (
+           Vector,
+           fromList,
+           length,
+           (!),
+           drop,
+           takeWhile,
+           findIndexR
+       )
+import Data.Vector.Primitive qualified as Primitive
+       (
+           Vector (Vector),
+           force,
+           null,
+           length,
+           take,
+           drop,
+           splitAt,
+       )
+import Database.LSMTree.Internal.Serialise
+       (
+           SerialisedKey (SerialisedKey')
+       )
+import Database.LSMTree.Internal.Entry
+       (
+           NumEntries (NumEntries)
+       )
+import Database.LSMTree.Internal.IndexCompact
+       (
+           PageNo (PageNo),
+           PageSpan (PageSpan)
+       )
+
+{-|
+    The version of the index and its serialisation format that is supported by
+    this module.
+-}
+supportedVersion :: Word32
+supportedVersion = 1
+
+{-|
+    A general-purpose fence pointer index.
+
+    An index is represented by a vector that maps the number of each page to the
+    key stored last in this page or, if the page is an overflow page, to the key
+    of the corresponding key–value pair. From this, the following constraints on
+    the vector follow:
+
+      * The vector must be non-empty.
+
+      * The elements of the vector must be non-decreasing.
```

Comment:
That the elements of the vector must be non-decreasing follows from the constraints on runs in conjunction with the way indexes are represented, but that the vector must be non-empty indeed doesn’t so clearly follow from anything: you could only argue that runs are supposed to be non-empty and the non-emptiness of the vector follows from this. You’re absolutely right, though, that the non-emptiness is required for `search` working properly. I’ll try to make the above documentation clearer.

Reply:
To be precise, it’s not just my particular implementation of `search` that doesn’t work with an empty vector, but no correct implementation of `search` can work with an empty vector, since an empty vector would represent a run without pages and there are no valid page spans to return for such a run.

Reply:
The improved documentation is in 4948b05b99ef7eac5f0a5260ab4aa4acdd2a39a5.